### ISSUE TITLE
Add `Graph` namespace

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#3906]
TBD

Define an API for a `Graph` namespace backed by Datalog:

```scala
///
/// Returns the transitive closure of the graph `g`.
///
def closure(g: List[(String, String)]): List[(String, String)] = ???

///
/// Returns the vertices that are reachable from the origin `o` in the graph `g`.
///
def reachable(o: String, g: List[(String, String)]): Set[String] = ???

///
/// Returns the vertices that are unreachable from the origin `o` in the graph `g`.
///
def unreachable(o: String, g: List[(String, String)]): Set[String] = ???

///
/// Returns `true` if there is a path from `src` to `dst` in the graph `g`.
///
def isConnected(src: String, dst: String, g: List[(String, String)]): Set[String] = ???

///
/// Returns the strongly connected components of the graph `g`.
///
def stronglyConnectedComponents(g: List[(String, String)]): List[Set[String]] = ???

///
/// Returns the topological sort of the graph `g`.
///
def topologicalSort(g: List[(String, String)]): Option[List[String]] = ???

///
/// ...
///
def isCyclic(g: ...): Bool = ???

///
/// Returns the shortest distance from `src` to `dst` in the graph `g`.
///
def distance(src: String, dst: String, g: List[(String, d, String)]): d with LatticeStuff[d] = ???

///
/// Returns the shortest distance from `o` to every other vertex in the graph `g`.
///
def distances(o: String, g: List[(String, d, String)]): Map[String, d] with LatticeStuff[d] = ???

///
/// Returns the given graph `g` as a GraphViz dot string.
///
def toGraphViz(g: List[(String, d, String)]): String = ???

///
/// Returns the undirected graph of the given directed graph.
///
def toUndirected(...)


/// Minimum spanning tree?
```

In the above examples I have used `List` and `String` but really this should work for any `m[t]` and `t`.

The distance functions probably require some more explanation, but the TLDR is that they require a lattice and a sum function. Maybe this requires a type class too, like:

```
class Distance[t] with AllTheLatticeClassesHere[t] {
  def add(x: t, y: t): t
}
```

Ideas:
- SCC?

We should use `m[(t, t)] with Foldable[m]` as the main representation of a graph (edges).

**Comment by magnus-madsen** [comment]
@JonathanStarup @mlutze @stephentetley @jaschdoc @ninaandrup 

Do you have ideas for additional functions that we would want?

Are any of you interested in implementing these functions (using Datalog of course)-- mostly addressed to Jakob and Nina.

**Comment by magnus-madsen** [comment]
https://hackage.haskell.org/package/algebraic-graphs-0.6/docs/Algebra-Graph-Class.html
https://hackage.haskell.org/package/containers-0.6.5.1/docs/Data-Graph.html

**Comment by stephentetley** [comment]
A `toDot` function to generate GraphViz files would be nice for small graphs. I thought I'd written one already but I can't find it. 

Datalog versions of transitive closure and reachability are listed in this paper:

https://www.redaktion.tu-berlin.de/fileadmin/fg131/Publikation/Papers/datalog-to-flink_2020.pdf

**Comment by stephentetley** [comment]
BTW I've got a wrapper module over Java's `java.lang.Processs` that would be useful in the stdlib (after some polish), then we could launch programs like GraphViz's `dot` from Flix.

**Comment by mlutze** [comment]
I guess we cannot actually get the path from datalog? But it would still be nice to have a shortest path function, and maybe some heuristic searches as well (A*, etc).

Topological sort would be handy as well.

**Comment by jaschdoc** [comment]
I think this would be fun to work on but I think I have tickets in my backlog that should be finished before I start something new so if Nina finds this interesting I think she should have it.

**Comment by ninaandrup** [comment]
I would be up for starting to implement this, if Jakob has too much on his plate for now :-)

**Comment by jaschdoc** [comment]
I'm not sure. I think Magnus wrote something in the June meeting issue

**Comment by magnus-madsen** [comment]
I was updating this issue.

**Comment by magnus-madsen** [comment]
We should also add all pairs shortest paths:

```
def apsp (g: List[(String, Int32 , String)]): List[(String, String, N)] =
    let db = inject g into Edge;
    let pr = #{
        Dist(x, x; N(0)) :- Edge(x, _, _).
        Dist(x, x; N(0)) :- Edge(_, _, x).

        Dist(x, y; N(d)) :- Edge(x, d, y).
        Dist(x, y; add(d1 , d2)) :- Dist(x, z; d1) , Dist(z, y; d2).
    };
    query db , pr select (x, y, d) from Dist (x, y; d)
```

**Comment by magnus-madsen** [comment]
@JonathanStarup Will you start looking into this a bit :)

(In the interest of the paper)

**Comment by magnus-madsen** [comment]
@JonathanStarup We could also consider adding Paul's n-degrees separation?

**Comment by magnus-madsen** [comment]
@JonathanStarup Anything missing or can I close?
